### PR TITLE
Nh/facebook & google auth

### PR DESCRIPTION
**Pull request by nhachicha**
This PR adds Google Sign-In and Facebook OAuth with ROS.

This has been successfully tested with `realm-object-server-bundled_node_darwin-1.0.0-BETA-2.1-g4a76af85`  https://ci.realm.io/job/realm/job/realm-sync-services/job/master/288/

you need first to modify the configuration of the server to include Facebook & Google as a provider 
in `object-server/configuration.yml` 

```
providers: 
    google:
      ## The client ID as retrieved when setting up the app in the Google
      ## Developer Console.
      clientId: '299391824375-k3hdrdtq8g4rkh1i0pvomn397ehk92nl.apps.googleusercontent.com'

    ## This enables authentication via a Facebook access token for a specific app.
    ## This provider needs no configuration (uncommenting the next line enables it).
     facebook: {}
```

You need also to provide a signed APK using this debug [keystore](https://www.dropbox.com/s/e6d4quorwct6z8u/debug.keystore?dl=0) (because that's the authorised SHA1 of the APK allowed with Facebook & Google console configuration)

Note: running the app from Android Studio will not use a signed APK (even for `Release` flavour)
you need to generate & install an APK from AS (_Build_ -> _Generate Signed APK_) or run the following gradle tasks
 ```gradle
./gradlew assembleRelease
adb install app/build/outputs/apk/app-release.apk
```

screen shot 2016-10-16 at 02 15 22

screen shot 2016-10-16 at 02 15 00

screen shot 2016-10-16 at 02 17 41

### CODE REVIEW

**Comment by nhachicha** [comment]
@realm/java

**Comment by zaki50** [comment]
How about adding the debug keystore into the repository?

File: RealmTasks Android/app/build.gradle [hunk 1]
```diff
@@ -29,6 +29,21 @@ android {
             buildConfigField "String", "OBJECT_SERVER_IP", "\"${host}\""
         }
     }
+
+    signingConfigs {
+        config {
+            storeFile file("${System.env.HOME}/.android/debug.keystore")
```
Comment (cmelchior) [inline RealmTasks Android/app/build.gradle]: Just adding the keystore to the repo seems easier?

File: RealmTasks Android/app/build.gradle [hunk 2]
```diff
@@ -44,4 +59,10 @@ dependencies {
     compile('io.realm:android-adapters:1.3.0') {
         exclude group: "io.realm"
     }
+    compile('com.facebook.android:facebook-android-sdk:4.15.0')
```
Comment (cmelchior) [inline RealmTasks Android/app/build.gradle]: Perhaps add `//Dependency for Facebook Sign-in`?

File: RealmTasks Android/app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -30,6 +33,13 @@
                 android:name="android.support.PARENT_ACTIVITY"
                 android:value=".TaskListActivity"/>
         </activity>
+
+        <activity android:name="com.facebook.FacebookActivity"
+            android:configChanges=
+                "keyboard|keyboardHidden|screenLayout|screenSize|orientation"
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/AndroidManifest.xml]: No need to wrap this?

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/RealmTasksApplication.java [hunk 4]
```diff
@@ -34,6 +36,7 @@
     public void onCreate() {
         super.onCreate();
         Realm.init(this);
+        FacebookSdk.sdkInitialize(getApplicationContext());
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/RealmTasksApplication.java]: `getApplicationContext()` -> `this`

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/RegisterActivity.java [hunk 5]
```diff
@@ -30,27 +31,41 @@
...
     private EditText passwordConfirmationView;
     private View progressView;
     private View registerFormView;
+    private FacebookAuth facebookAuth;
+    private GoogleAuth googleAuth;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/RegisterActivity.java]: Why this?

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+import com.facebook.CallbackManager;
+import com.facebook.FacebookCallback;
+import com.facebook.FacebookException;
+import com.facebook.login.LoginResult;
+import com.facebook.login.widget.LoginButton;
+
+/**
+ * Provide authentication using users Facebook account.
+ */
+
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java]: Remove whitespace

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import android.support.v4.app.FragmentActivity;
+
+import com.facebook.CallbackManager;
+import com.facebook.FacebookCallback;
+import com.facebook.FacebookException;
+import com.facebook.login.LoginResult;
+import com.facebook.login.widget.LoginButton;
+
+/**
+ * Provide authentication using users Facebook account.
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java]: Is there some reference Facebook docs we can point to?

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+import com.google.android.gms.common.ConnectionResult;
+import com.google.android.gms.common.SignInButton;
+import com.google.android.gms.common.api.GoogleApiClient;
+
+import io.realm.realmtasks.R;
+
+/**
+ * * Provide authentication using users Google account registered with the device.
+ */
+
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java]: Remove whitespace

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
+import com.google.android.gms.auth.api.signin.GoogleSignInResult;
+import com.google.android.gms.common.ConnectionResult;
+import com.google.android.gms.common.SignInButton;
+import com.google.android.gms.common.api.GoogleApiClient;
+
+import io.realm.realmtasks.R;
+
+/**
+ * * Provide authentication using users Google account registered with the device.
```
Comment (cmelchior) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java]: Is there some reference docs we can point to?

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/RegisterActivity.java (hunk 5 above)
Comment (nhachicha) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/RegisterActivity.java]: Was testing Google Sign-In failure. Will remove now.

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java (hunk 7 above)
Comment (nhachicha) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/facebook/FacebookAuth.java]: will add this https://developers.facebook.com/docs/facebook-login/android

File: RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java (hunk 9 above)
Comment (nhachicha) [inline RealmTasks Android/app/src/main/java/io/realm/realmtasks/auth/google/GoogleAuth.java]: Will add this https://developers.google.com/identity/sign-in/android/start-integrating

**Comment by jpsim** [comment]
We should probably do the same on the iOS/Mac side once this lands. I know @TimOliver had made some progress previously in #61, but that's pretty outdated by now.

**Comment by nhachicha** [comment]
@realm/java updated

File: RealmTasks Android/app/build.gradle [hunk 10]
```diff
@@ -44,4 +59,11 @@ dependencies {
     compile('io.realm:android-adapters:1.3.0') {
         exclude group: "io.realm"
     }
+    // Dependency for Facebook Sign-in
+    compile('com.facebook.android:facebook-android-sdk:4.15.0')
```
Comment (dalinaum) [inline RealmTasks Android/app/build.gradle]: This line includes parenthesis and next line doesn't include parenthesis. Should we remove parenthesis here or add parenthesis on next line?

**Pull request merged**